tests/templates/test_subroutines.py

```diff
@@ -322,6 +326,120 @@ def circuit(theta, phi, varphi):
+                   [13, qml.CNOT, [1,2], []], [14, qml.CNOT, [2,3], []], [15, qml.CNOT, [3,4], []],
+                   [17, qml.CNOT, [3,4], []], [18, qml.CNOT, [2,3], []], [19, qml.CNOT, [1,2], []]]
+                   ),
+
+        ([10,11], [[2 , qml.CNOT, [10,11] , []], [4  , qml.CNOT, [10,11], []],
+                   [9 , qml.CNOT, [10,11] , []], [11 , qml.CNOT, [10,11], []]]
+                   )        
+        ]
+    )
+    def test_single_ex_op_operations(self, ph, ref_gates):
+        """Test the correctness of the SingleExcitationOp template including the gate count
+        and order, the wires each operation acts on and the correct use of parameters 
+        in the circuit."""
+
+        sqg = 10
+        cnots = 4*(ph[1]-ph[0])
+        weight = np.pi/3
+        with qml.utils.OperationRecorder() as rec:
+            SingleExcitationOp(weight, wires=ph)
+
+        idx = ref_gates[0][0]
+
+        exp_gate = ref_gates[0][1]
+        res_gate = rec.queue[idx]
+
+        exp_wires = ref_gates[0][2]
+        res_wires = rec.queue[idx]._wires
+
+        exp_weight = ref_gates[0][3]
+        res_weight = rec.queue[idx].parameters        
+
+        assert len(rec.queue) == sqg + cnots
+        assert isinstance(res_gate, exp_gate) 
+        assert res_wires == exp_wires
+        assert res_weight == exp_weight
+
+    @pytest.mark.parametrize(
+        ("weight", "ph", "msg_match"),
+        [
+            ( 0.2      , [0]         , "'wires' must be of shape"),
+            ( 0.2      , []          , "'wires' must be of shape"),
+            ([0.2, 1.1], [0,2]       , "'weight' must be of shape"),
+            ( 0.2      , None        , "wires must be a positive integer"),
+            ( 0.2      , ["a", "b"]  , "wires must be a positive integer"),
+            ( 0.2      , [1.13, 5.23], "wires must be a positive integer"),
+            ( 0.2      , [3, 3]      , "wires_1 must be > wires_0"),
+            ( 0.2      , [3, 1]      , "wires_1 must be > wires_0")
+        ]
+    )
+    def test_single_excitation_op_exceptions(self, weight, ph, msg_match):
+        """Test that SingleExcitationOp throws an exception if ``weight`` or 
+        ``ph`` parameter has illegal shapes, types or values."""
+        dev = qml.device("default.qubit", wires=5)
+
+        def circuit(weight=weight, ph=ph):
+            SingleExcitationOp(weight=weight, wires=ph)
+            return qml.expval(qml.PauliZ(0))
+
+        qnode = qml.QNode(circuit, dev)
+
+        with pytest.raises(ValueError, match=msg_match):
+            qnode(weight=weight, ph=ph)
+
+    @pytest.mark.parametrize(
+        ("weight", "ph", "expected"),
+        [
+            ( 2.21375586 , [0, 2], [-0.59956665, 1.        , 0.59956665, -1.]),
+            ( -5.93892805, [1, 3], [ 1.        , 0.94132639, -1.       , -0.94132639])
+        ]
+    )
+    def test_integration(self, weight, ph, expected, tol):
```

Comment:
Do you also want to add this template to the ``test_integration`` tests? Instructions are in the docstring of the test file...

Reply:
@mariaschuld , I have tried this but the test is not passing.
Based on the docstring I extend ``QUBIT_DIFFABLE_NONDIFFABLE`` adding 
```
(qml.templates.SingleExcitationUnitary,
                               {'weight': 0.56},
                               {'wires': [0, 1]}),
```
The error message raised by ``get_shape``:
```
inpt = range(0, 2)

    def get_shape(inpt):
        """Turn ``inpt`` into an array and return its shape.
    
        Args:
            inpt (scalar, list or array): input to a qnode
    
        Returns:
            tuple: shape of ``inpt``
        """
    
        # avoids incorrect assignment of shape
        if isinstance(inpt, (float, int, complex)):
            shape = ()
    
        else:
            # turn lists into array to get shape
            if isinstance(inpt, list):
                inpt = np.array(inpt)
    
            try:
                shape = inpt.shape
            except AttributeError:
>               raise ValueError("could not extract shape of object of type {}".format(type(inpt)))
E               ValueError: could not extract shape of object of type <class 'range'>
```
I can see this line  ``all_args['wires'] = range(2)`` inside ``test_qubit_qnode_primary_args.circuit`` function which might be causing the problem. Notice I use ``wires`` as keyword in the template. Probably something is being overwriten in ``all_args`` ??

Reply:
You are right, the way you use the check functions is fully ok, but uncovered a few things that need to be adjusted:

* the integration test was creating ranges a wire arguments, and while the other templates accept a range as well yours does not
* the ``check_shape`` test forgot to check for scalars of type Variable, because I never used them elsewhere

I fixed those two things and added your template to the integration test!